### PR TITLE
Edit Unavailability Patch Endpoint #FE-58

### PR DESCRIPTION
**Pull request by TaiHaDev**
## Changes
Created a Patch Endpoint `volunteer_unavailability` with a function `edit_event` to be later put in repository folder to access and edit unavailability in the database. The endpoint is manually tested as it is a learning task:
image

### CODE REVIEW

File: controllers/v2/unavailability/volunteer_unavailability_v2.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        return True
+    except Exception as e:
+        session.rollback()
+        logging.error(e)
+        return False
+
+
+volunteer_unavailability_v2_bp = Blueprint('volunteer_unavailability_v2', __name__)
+api = Api(volunteer_unavailability_v2_bp, "/unavailability_v2")
+api.add_resource(volunteer_unavailability, "/editUnavailableEvent")
```
Comment (BenMMcLean) [inline controllers/v2/unavailability/volunteer_unavailability_v2.py:54]: These endpoint names don't follow standard [REST endpoint naming conventions](https://restfulapi.net/resource-naming/)

File: controllers/v2/unavailability/volunteer_unavailability_v2.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        args = edit_parser.parse_args()
+        with session_scope() as session:
+            success = edit_event(session, **args)
+            if success:
+                return jsonify({"success": True, "message": "updated successfully"})
+            else:
+                return jsonify({"success": False, "message": "Failed to update"})
+
+
+def edit_event(session, userId, eventId, title=None, start=None, end=None, periodicity=None):
```
Comment (BenMMcLean) [inline controllers/v2/unavailability/volunteer_unavailability_v2.py:32]: This is an inappropriate place to put data operations. Please move it to a repository.

File: controllers/v2/unavailability/volunteer_unavailability_v2.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+edit_parser.add_argument("periodicity", type=int)
+
+
+class volunteer_unavailability(Resource):
+    def patch(self):
+        args = edit_parser.parse_args()
+        with session_scope() as session:
+            success = edit_event(session, **args)
+            if success:
+                return jsonify({"success": True, "message": "updated successfully"})
```
Comment (BenMMcLean) [inline controllers/v2/unavailability/volunteer_unavailability_v2.py:27]: Just change the HTTP status code depending on the outcome of the operation

File: controllers/v2/unavailability/volunteer_unavailability_v2.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+edit_parser.add_argument("start", type=inputs.datetime_from_iso8601)
+edit_parser.add_argument("end", type=inputs.datetime_from_iso8601)
+edit_parser.add_argument("periodicity", type=int)
+
+
+class volunteer_unavailability(Resource):
+    def patch(self):
+        args = edit_parser.parse_args()
+        with session_scope() as session:
+            success = edit_event(session, **args)
```
Comment (BenMMcLean) [inline controllers/v2/unavailability/volunteer_unavailability_v2.py:25]: Surely you should check the event exists before attempting to edit it? If it doesn't return a 400.

**Review by BenMMcLean — CHANGES_REQUESTED** [review]
Couple points to address + in the future, please name your branches using the format username/<feature, bugfix, hotfix, fix, etc>/name. Our pipelines are expecting this format and use it for versioning

File: repository/volunteer_unavailability_v2.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+            event.title = title
+        if start is not None:
+            event.start = start
+        if end is not None:
+            event.end = end
+        if end is not None:
+            event.periodicity = periodicity
+        session.commit()
+        return True
+    except Exception as e:
```
Comment (BenMMcLean) [inline repository/volunteer_unavailability_v2.py:22]: It's good enough (I'll approve your PR), but I generally prefer to propagate exceptions up to the presentation layer (in this case the endpoint) and handle it there. It just offers the developer more clarity and flexibility. As such, I would have caught it, rolled back, and then thrown `e`

**Review by BenMMcLean — APPROVED** [review]

**Comment by BenMMcLean** [comment]
Good work!

**Comment by TaiHaDev** [comment]
Hi Ben, thank you so much for your valuable feedback! I just fixed it and pushed it again hopefully you can have a look at it :-) I put the data operation code in the wrong place because it was a learning task to show understanding towards the project so I didn't do it carefully but the rest of the feedback is important for me :-).
These pull requests from the team are for you to see each member's understanding of the project and receive valuable feedback like the above before actual development. We will work on these endpoints for the next audit and make them much better.

**Comment by TaiHaDev** [comment]
> Couple points to address + in the future, please name your branches using the format username/<feature, bugfix, hotfix, fix, etc>/name. Our pipelines are expecting this format and use it for versioning

will it be something like this: john/edit-unavailability ?

**Comment by BenMMcLean** [comment]
> > Couple points to address + in the future, please name your branches using the format username/<feature, bugfix, hotfix, fix, etc>/name. Our pipelines are expecting this format and use it for versioning
> 
> will it be something like this: john/edit-unavailability ?

It would be something like john/feature/edit-unavailability

**Review by BenMMcLean — APPROVED** [review]

**Pull request merged**